### PR TITLE
Support conditions for transactional operations

### PR DESCRIPTION
**Pull request by Torch3333**
**Context** 
Setting conditions for Put or Delete in a transaction was not supported. Setting conditions in the operation builder was possible, but they were ignored.

**Changes**
Setting condition is now fully supported on Put and Delete operation for consensus commit (regular and 2pc) or Jdbc transactions. For example :

https://github.com/scalar-labs/scalardb/blob/bdf51bd170101939266956a340734f0fe37d300c/docs/api-guide.md?plain=1#L594-L612

### CODE REVIEW

**Comment by komamitsu** [comment]
@Torch3333 Could you tell me a bit more detail of the use case?

I first thought this PR was an enhancement of current ID and VERSION based conflict check. But ID and VERSION based conflict check is already enough to detect conflict operations. So, on second thought, I guess the use case for this PR is to make the following pseudo code shorter. But I'm not sure if my guess is correct...

old
```
  account = get(k)
  if (account != null && account.balance == 500) {
    account.balance -= 200
    put(account)
  }
```

new
```
  put(account, condition: { balance == 500 })
```

**Comment by Torch3333** [comment]
@komamitsu 
Yes, It does not relate to conflict checks. 
When the condition is not validated, an exception is thrown (`CommitConflictException` for consensus commit transaction, `PreparationConflictException` for 2pc, and `CrudException` for jdbc transaction) so the usage is more appropriate for validating constraints.

old
```
account = get(k)
if (account != null && account.active == true && account.balance >=200) {
  account.balance -= 200
  put(account)
} else {
  throw new Exception("constraint violated")
}
```

new
```
account = get(k)
if (account != null) {
  account.balance -= 200
  put(account, {condition : {account.active == true, account.balance >= 200}})
}
```
Does this answer your question?

**Comment by komamitsu** [comment]
@Torch3333 Thanks for the example! I understand it's like constraint validation.

So, I'm wondering how users can address when ConflictException is thrown. With current implementation, I think ConflictException means like "Your transaction is conflicted with other transaction. Please retry it so that it would succeed." But with this PR change, users can face a ConflictException that occurs due to constraint violation and retry may not work for the cause...? For instance, in the above case of `account.balance >= 200`, users may keep retrying the code if the original account's balance is less than 200.

**Review by brfrn169 — COMMENTED** [review]
LGTM overall. Left some minor comments. Since it's core logic, I'll take a look at it again later. Thanks.

File: core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransaction.java [hunk 1]
```diff
@@ -81,14 +81,10 @@ public List<Result> scan(Scan scan) throws CrudException {
...
-    // Ignore the condition in the put
-    if (put.getCondition().isPresent()) {
-      logger.warn("ignoring the condition of the mutation: {}", put);
-      put.withCondition(null);
-    }
-
     try {
-      jdbcService.put(put, connection);
+      if (!jdbcService.put(put, connection)) {
+        throw new CrudException("no mutation was applied", txId);
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransaction.java:86]: Ok. I will update with the exception type. For the message, I will add`put` in it to be more precise: `the put operation condition is not satisfied

File: core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransaction.java [hunk 2]
```diff
@@ -108,14 +104,10 @@ public void put(List<Put> puts) throws CrudException {
...
-    // Ignore the condition in the delete
-    if (delete.getCondition().isPresent()) {
-      logger.warn("ignoring the condition of the mutation: {}", delete);
-      delete.withCondition(null);
-    }
-
     try {
-      jdbcService.delete(delete, connection);
+      if (!jdbcService.delete(delete, connection)) {
+        throw new CrudException("no mutation was applied", txId);
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransaction.java:109]: Ok. I will update with the exception type. For the message, I will add `delete` in it to be more precise: `the delete operation condition is not satisfied`

**Comment by feeblefakie** [comment]
https://github.com/scalar-labs/scalardb/pull/837#issuecomment-1549096971

I think this is a really good point.
If we cannot distinguish conflicts from condition mismatch, it's gonna be an issue.

**Comment by Torch3333** [comment]
Thank you all for the feedback. 

After discussing with @feeblefakie and @brfrn169, we are thinking of adding logic to verify the user condition for `PutIf` and `DeleteIf` in the `PrepareMutationComposer` and not pass the condition to the storage layer to validate the condition. 
Additionally, we want to throw a dedicated exception when the user conditions are not verified.
I will switch back the PR to draft while implementing the changes.

**Comment by Torch3333** [comment]
I created a new PR https://github.com/scalar-labs/scalardb/pull/899#pullrequestreview-1469012428 since most of the code has been reworked.

**Pull request closed without merging**